independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/PooledBufferAllocator.java

```diff
@@ -0,0 +1,107 @@
+package io.quarkus.bootstrap.runner;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A thread-safe buffer allocator that pools heap buffers.<br>
+ * It's currently implemented using a concurrent linked stack, but the users shouldn't rely on how it's implemented,
+ * as it's subject to change. Conversely, users can rely on the fact that it's thread-safe.
+ *
+ * Users which relies on its buffer API, but doesn't need anymore to pool buffers, can call {@link #disablePooling()}
+ * which will make it behave as if it was disabled from the start and release any buffer that was still pooled.
+ */
+public final class PooledBufferAllocator {
+
+    // This is a sentinel value used to forcibly disable pooling.
+    private static final Buffer EMPTY = new Buffer();
+
+    /**
+     * A buffer that contains the data of a resource.<br>
+     * It can be released back (once) to its allocator via {@link PooledBufferAllocator#release(Buffer)}.
+     */
+    public static final class Buffer {
+        private byte[] data;
+        private int length;
+        private Buffer next;
+
+        private Buffer(byte[] data, int length) {
+            assert data != null && length >= 0;
+            this.data = data;
+            this.length = length;
+        }
+
+        private Buffer() {
+        }
+
+        public byte[] array() {
+            return data;
+        }
+
+        public int length() {
+            return length;
+        }
+
+        /**
+         * It moves the data from this buffer to a new one, and returns the new one.
+         */
+        private Buffer moveDataToNewBufferOf(final int requiredCapacity) {
+            byte[] data = this.data;
+            // it's forbidden to modify EMPTY
+            if (this != EMPTY) {
+                this.data = null;
+            }
+            if (data == null || data.length < requiredCapacity) {
+                data = new byte[requiredCapacity];
```

Comment:
Did you consider that you might get stronger allocation savings by oversizing a little here?
As long as `requiredCapacity` isn't a large number, you could e.g. double the figure and get much higher likelyhood of actual reuse.

Reply:
Also: I feel there's some buffers being wasted here? You're throwing away a perfectly fine data array if it's too small for the current request. Shouldn't it be returned to the pool instead?

Totally agree that it needs to remain a relatively simple algorythm, and also that we don't want to spend the cost of tracking many small buffers which are in practice too small to be reused effectively... so perhaps rather than returning them to the pool you really should be looking at over-allocating instead, to make this case unlikely.

Reply:
It's a good point, just, differently from an array list, when we grow it's usually for more than a single byte increase and allocations are past KB: doubling (or +50%) can mean some data, but I agree...
Maybe I can run the basic config, collect the biggest allocation there and presize the pool with it. It's an educated bet which solve the smallest allocations in one go (although is prone to be invalid in a future and I have no idea how to write a test for it).

Reply:
In the current version I'm not fixing this yet, because:
- presizing the buffer still make it cost even if not used (which is unlikely because we know some class will be loaded and most of the time such are *at least* the ones of hello world-like startup examples)
- enlarging by X should be based on the actual size, which is not a simple heuristic: we don't want to "always"  allocate 50% more data cause adding 1/2 of 20 MB isn't the same as adding 1/2 of 100 K: leaving the classes to do it instead will save deciding and still getting a decent although not optimizal saving, given that the allocations won't be pooled at all in the "before" branch